File: PrimaMateria/PrimaMateria/Classes/XTRPeriodicTableViewController.m
//
//  XTRFirstViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright 2016 xTrensa. All rights reserved.
//

#import "PrimaMateria.h"

@interface XTRPeriodicTableViewController ()
- (void)showPopupForButton: (id) sender;
- (void)showElementPanelForElementAtIndex: (int) anIndex;
- (void)setupPopUp;
- (void)toggleMolecularCalculatorState: (BOOL) aFlag;
@end

@implementation XTRPeriodicTableViewController

#pragma mark Private Methods

- (void)showPopupForButton: (id) sender {
    [XTRPropertiesStore storeViewTitle: self.title];
    [XTRPropertiesStore storeAtomicNumber: @([sender tag])];

    CGRect aRect = CGRectMake(10, 10, 15, 15);
    self.elementBalloonViewController.popoverPresentationController.sourceRect = aRect;
    self.elementBalloonViewController.popoverPresentationController.sourceView = sender;
    [self presentViewController: self.elementBalloonViewController animated:YES completion:nil];
}

- (void)showElementPanelForElementAtIndex: (int) anIndex {
    [XTRPropertiesStore storeViewTitle: self.title];
    [XTRPropertiesStore storeAtomicNumber:@(anIndex)];
    [self performSegueWithIdentifier: SHOW_INSPECTOR_FROM_PERIODIC_TABLE sender: self];
}

- (void)setupPopUp {
	self.elementBalloonViewController = [[XTRAppDelegate storyboard] instantiateViewControllerWithIdentifier: NSStringFromClass([XTRElementBalloonViewController class])];
    CGSize contentSize = CGSizeMake(324.0f, 210.0f);
    self.elementBalloonViewController.preferredContentSize = contentSize;
    self.elementBalloonViewController.modalPresentationStyle = UIModalPresentationPopover;
}

- (void)toggleMolecularCalculatorState: (BOOL) aFlag {
    if (!aFlag) {
        self.molecularCalculatorState = NO;
        [self.molecularCalculatorViewController.view removeFromSuperview];
        [self.molecularCalculatorViewController viewWillDisappear: YES];
    } else {
        self.molecularCalculatorState = YES;
        [self.view addSubview: self.molecularCalculatorViewController.view];
    }
}

#pragma mark - Action Methods

- (IBAction) showMolecularCalculator: (UISwitch *) sender {
    [self toggleMolecularCalculatorState: sender.on];
}

- (IBAction) showElementInspector: (id) sender {
    if (self.molecularCalculatorState) {
        XTRElement *element = [[XTRDataSource sharedInstance] elementAtIndex:[sender tag]];
        [self.molecularCalculatorViewController setElement: element];
    } else {
        BOOL defaultState = [XTRPropertiesStore retreiveElementBubblesState];
        
        if (defaultState) {
            [self showPopupForButton: sender];
        } else {
            [self showElementPanelForElementAtIndex:[sender tag]];
        }
    }
}

#pragma mark - Notification Methods

-(void)closeBubble: (NSNotification *)aNotification {
    [self dismissViewControllerAnimated: YES completion: nil];
}

#pragma mark - View Management Methods

- (void)viewDidLoad {
    [super viewDidLoad];
	self.molecularCalculatorViewController = [[XTRAppDelegate storyboard] instantiateViewControllerWithIdentifier: NSStringFromClass([XTRMolecularCalculatorViewController class])];
    self.molecularCalculatorViewController.view.frame = self.swapView.frame;
    [self.swapView removeFromSuperview];
    [self setupPopUp];
    [self toggleMolecularCalculatorState: NO];
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(closeBubble:) name: NOTIFICATION_INSPECTOR_DISMISSED object: nil];

}

- (BOOL) shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation)interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

#pragma mark - Memory Management Methods

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
    self.swapView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRPeriodicTableViewController.swift
//
//  XTRPeriodicTableViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/15/16.
//  Copyright © 2016 xTrensa. All rights reserved.
//

@objc class XTRPeriodicTableViewController : UIViewController  {
    @IBOutlet var swapView : UIView!
    
    var molecularCalculatorState : Bool?
    var elementBalloonViewController : XTRElementBalloonViewController?
    var molecularCalculatorViewController : XTRMolecularCalculatorViewController?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    func showElementPanelForElementAtIndex(anIndex: UInt) {
        XTRPropertiesStore.storeViewTitle(self.title)
        XTRPropertiesStore.storeAtomicNumber(anIndex)
        self.performSegueWithIdentifier(SHOW_INSPECTOR_FROM_PERIODIC_TABLE, sender: self)
    }
    
    func setupPopUp() {
        self.elementBalloonViewController = XTRAppDelegate.storyboard().instantiateViewControllerWithIdentifier("XTRElementBalloonViewController") as? XTRElementBalloonViewController
        let contentSize : CGSize = CGSizeMake(324.0, 210.0)
        self.elementBalloonViewController!.preferredContentSize = contentSize
        self.elementBalloonViewController!.modalPresentationStyle = UIModalPresentationStyle.Popover
    }
    
    func toggleMolecularCalculatorState(aFlag: Bool) {
        if (!aFlag) {
            self.molecularCalculatorState = false
            self.molecularCalculatorViewController!.view.removeFromSuperview()
            self.molecularCalculatorViewController!.viewWillDisappear(true)
        } else {
            self.molecularCalculatorState = true
            self.view.addSubview(self.molecularCalculatorViewController!.view)
        }
    }
    
    // MARK: - Action Methods
    
    @IBAction func showPopupForButton(sender: UIButton) {
        XTRPropertiesStore.storeViewTitle(self.title)
        XTRPropertiesStore.storeAtomicNumber(sender.tag)
        
        let aRect: CGRect = CGRectMake(10, 10, 15, 15)
        self.elementBalloonViewController!.popoverPresentationController!.sourceRect = aRect
        self.elementBalloonViewController!.popoverPresentationController!.sourceView = sender
        self.presentViewController(self.elementBalloonViewController!, animated:true, completion:nil)
    }
    
    @IBAction func showMolecularCalculator(sender: UISwitch) {
        self.toggleMolecularCalculatorState(sender.on)
    }
    
    @IBAction func showElementInspector(sender: UIButton) {
        if (self.molecularCalculatorState!.boolValue) {
            let element : XTRElement = XTRDataSource.sharedInstance().elementAtIndex(UInt(sender.tag))
            self.molecularCalculatorViewController!.setElement(element)
        } else {
            let defaultState : Bool? = XTRPropertiesStore.retreiveElementBubblesState()
            
            if (defaultState!.boolValue) {
                self.showPopupForButton(sender)
            } else {
                self.showElementPanelForElementAtIndex(UInt(sender.tag))
            }
        }
    }
    
    // MARK: mark - Notification Methods
    
    func closeBubble(aNotification: NSNotification) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.molecularCalculatorViewController = XTRAppDelegate.storyboard().instantiateViewControllerWithIdentifier("XTRMolecularCalculatorViewController") as? XTRMolecularCalculatorViewController
        self.molecularCalculatorViewController!.view.frame = self.swapView.frame
        self.swapView.removeFromSuperview()
        self.setupPopUp()
        self.toggleMolecularCalculatorState(false)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(XTRPeriodicTableViewController.closeBubble(_:)), name: NOTIFICATION_INSPECTOR_DISMISSED, object: nil)
    }
    
    override func shouldAutorotate() -> Bool {
        return true
    }
    
    override func supportedInterfaceOrientations() -> UIInterfaceOrientationMask {
        return [UIInterfaceOrientationMask.LandscapeLeft, UIInterfaceOrientationMask.LandscapeRight]
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self, name: NOTIFICATION_INSPECTOR_DISMISSED, object: nil)
        self.swapView = nil
    }
}
